sycl/source/detail/device_image_impl.hpp

```diff
@@ -38,6 +42,193 @@ template <class T> struct LessByHash {
   }
 };
 
+namespace syclex = sycl::ext::oneapi::experimental;
+
+using include_pairs_t =
+    std::vector<std::pair<std::string /* name */, std::string /* content */>>;
+
+// Bits representing the origin of a given image, i.e. regular offline SYCL
+// compilation, interop, kernel_compiler online compilation, etc.
+constexpr uint8_t ImageOriginSYCLOffline = 1;
+constexpr uint8_t ImageOriginInterop = 1 << 1;
+constexpr uint8_t ImageOriginKernelCompiler = 1 << 2;
+
+// Helper class to track and unregister shared SYCL device_globals.
```

Comment:
Why do we need these two helper classes at all? Couldn't underlying sycl objects track themselves? The helper would really benefit from more comments...

Reply:
This is only used for RTC binaries, which mean there are no actual host-side `device_global` objects. As such, there would be no owners for the `shared_ptr`. It is all contained to the `kernel_bundle`s created from the RTC compiled binaries.